uberfire-structure/uberfire-structure-backend/src/main/java/org/guvnor/structure/backend/repositories/ChangeRequestServiceImpl.java

```diff
@@ -1333,4 +1334,61 @@ JGitPathImpl createJGitPathImpl(final JGitFileSystem fs,
                                    objectId,
                                    isRealPath);
     }
-}
\ No newline at end of file
+
+    @Override
+    public List<ChangeRequestCommit> getCommits(final String spaceName,
+                                                final String repositoryAlias,
+                                                final Long changeRequestId) {
+
+        final Repository repository = resolveRepository(spaceName,
+                                                        repositoryAlias);
+
+        final ChangeRequest changeRequest = getChangeRequestById(spaceName,
+                                                                 repositoryAlias,
+                                                                 false,
+                                                                 changeRequestId);
+
+        final Git git = getGitFromBranch(repository,
+                                         changeRequest.getSourceBranch());
+
+        final String endCommitId = git.getLastCommit(changeRequest.getSourceBranch())
+                                      .getName();
+
+        final String startCommitId = changeRequest.getStartCommitId();
+
+        return git.listCommits(startCommitId,
+                               endCommitId)
+                  .stream()
+                  .map(c -> new ChangeRequestCommit(c.getName(),
+                                                    c.getFullMessage()))
+                  .collect(Collectors.toList());
+    }
+
+    @Override
+    public Boolean squashChangeRequest(final String spaceName,
+                                    final String repositoryAlias,
+                                    final Long changeRequestId,
+                                    final String commitMessage) {
+
+        final Repository repository = resolveRepository(spaceName,
+                                                        repositoryAlias);
+
+        final ChangeRequest changeRequest = getChangeRequestById(spaceName,
+                                                                 repositoryAlias,
+                                                                 false,
+                                                                 changeRequestId);
+
+        final Git git = getGitFromBranch(repository,
+                                         changeRequest.getSourceBranch());
+
+        final String startCommit = changeRequest.getStartCommitId();
+
+        git.squash(changeRequest.getSourceBranch(),
```

Comment:
Is it ok to squash the source branch? I mean, the whole history will be lost.

Reply:
I don't know, there is no spec

Reply:
Not sure if the version history is useful for users after the merge. If so, then we can merge first into the target branch and then squash the commits there (target branch).
@paulovmr ^

Reply:
@gabriel376 @caponetto I think it's better to keep the history, so the squash could be done in the target branch.

Reply:
Ok!

Reply:
Done